When entering an exception, increment reg_eipc if the cpu is halted

jwestfall69 opened the pull request:
If the cpu is in a halted state and an interrupt/exception occurs, the
saved reg_eipc should be set to the instruction after the halt.

yupferris reviewed (changes requested):
Thanks for this! A couple things:

Right now halt's implementation is kindof a hack. We basically just don't update the next PC and have the CPU execute the `halt` instruction over and over, which works generally, but is not very clean and certainly not very accurate from the rest of the system's POV.

Since this commit introduces a `halt` state, I think we should also use that to adjust the CPU `step` fn so that if we're in a halted state, we don't execute any instructions, but just run a single cycle and do nothing.

nit: I think we should call `halt` `is_halted` instead to make it more clear it's a boolean state (`halt` to me sounds like the name of a void fn that will halt the cpu).

jwestfall69 committed:
implement cpu halting

- When the cpu is halted make cpu.cycles() short circuit
- When an interrupt/exception happens while in a halted state make
reg_eipc point to the instruction after the halt

yupferris commented:
Looks great, thanks again! :)
